File: old/JNibbles/src/main/java/org/lagash/nibbles/Snake.java
/**
 * Copyright (C) 2017 DiegoG.
 */
package org.lagash.nibbles;

import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;

import org.newdawn.slick.Color;

public class Snake extends Vector implements ISnake {
    private static byte Count = 1;

    private int ticks = 0;

    private int id;

    private Color color;

    private int length;

    private List<Vector> trail;

    private ISnakeBehavior snake;

    public ISnakeBehavior getBehavior() {
        return this.snake;
    }

    public void setBehavior(ISnakeBehavior value) {
        this.snake = value;
    }
    
    public List<IVector> getTrail() {
        List<? extends IVector> list = trail;
        return Collections.unmodifiableList(list);
    }

    public int getTicks() { 
        return ticks; 
    }

    public Snake() {
        this.id = Count++;
        this.trail = new ArrayList<Vector>();
        this.length = 1;
    }

    public int getId() {
        return id;
    }

    public void setId(int value) {
        this.id = value;
    }

    public Color getColor() {
        return this.color;
    }

    public void setColor(Color value) {
        this.color = value;
    }
    public int getLength() {
        return length;
    }

    public void setLength(int value) {
        this.length = value;
    }

    public void changeDirection(Direction direction) {
        if(getDirection() != direction || getTrail().size() == 0) {
            Vector vec = new Vector();
            vec.setX( this.getX() );
            vec.setY( this.getY() );
            vec.setDirection( direction );
            trail.add(vec);
            setDirection( direction );
        }
    }

    public Vector reduceLength(int count) {
        length -= count;

        Vector first = trail.get(0);
        Vector second = trail.size() > 1 ? trail.get(1) : null;
        IPosition newFirst = first.moveNew(first.getDirection());
        first.setPosition(newFirst);
        if(second != null && 
            first.getX() == second.getX() && first.getY() == second.getY()) {
            trail.remove(first);
        }
        return first;
    }

    public void step() {
        this.ticks++;
        this.length++;
        this.move();
    }

    public Vector getTailPosition() {
        Vector first = trail.get(0);
        Vector newVec = new Vector();
        newVec.setX( first.getX() );
        newVec.setY( first.getY() );
        newVec.setDirection( first.getDirection() );
        return newVec;
    }

    public boolean willHitNextStep(Space space) {
        IPosition pos = moveNew(getDirection());
        if(pos.isValid(space.getTopX(), space.getTopY())
            && (space.get(pos.getX(), pos.getY()) == 0)) {
            return false;
        }
        return true;
    }

    public int willHitNextStepInfo(Space space) {
        IPosition pos = moveNew(getDirection());
        if(!pos.isValid(space.getTopX(), space.getTopY())) {
            return -1;
        }
        return space.get(pos.getX(), pos.getY());
    }

    public void printSnake() {
        System.out.println(String.format("  %d:(%d): x:%d, y:%d, %s, l:%d", id, ticks, getX(), getY(), getDirection(), getLength()));
        Iterator<Vector> iter = trail.iterator();
        while (iter.hasNext()) {
            Vector vec = iter.next();
            System.out.println(String.format("\t%d, %d %s", vec.getX(), vec.getY(), vec.getDirection()));
        }
    }
}

File: frontend/src/client/Snake.ts
import {Vector} from './Vector';
import {Space} from './Space';
import {ISnakeBehavior} from './ISnakeBehavior';
import {Direction} from './Direction';

export class Snake extends Vector {
    id: number;
    ticks: number;
    name: string;
    trail: Array<Vector>;
    length: number;
    color: string;
    behavior: ISnakeBehavior;

    constructor(id: number) {
        super();
        this.trail = new Array<Vector>();
        this.id = id;
        this.name = '🤦';
    }

    step() {
        this.ticks++;
        this.length++;
        this.move();
    }

    reduceLength(count: number): Vector {
        this.length -= count;

        let first = this.trail[0];
        let second = this.trail.length > 1 ? this.trail[1] : null;
        let newFirst = first.moveNewDirection(first.direction);
        first.x = newFirst.x;
        first.y = newFirst.y;
        if( second != null && 
            first.x === second.x && first.y === second.y) {
            this.trail.splice(0,1);
        }
        return first;
    }

    getTailPosition(): Vector {
        let first = this.trail[0];
        let newVec = new Vector();
        newVec.x = first.x;
        newVec.y = first.y;
        newVec.direction = first.direction;
        return newVec;
    }

    willHitNextStep(space:Space):boolean {
        let pos = this.moveNewDirection(this.direction);
        if(pos.isValidInBounds(space)
            && (space.map[pos.x][pos.y] === space.EMPTY)) {
            return false;
        }
        return true;
    }

    willHitNextStepInfo(space: Space) : number {
        let pos = this.moveNewDirection(this.direction);
        if(!pos.isValidInBounds(space)) {
            return -1;
        }
        return space.map[pos.x][pos.y];
    }

    changeDirection(direction: Direction) {
        if(this.direction !== direction || this.trail.length === 0) {
            let vec = new Vector();
            vec.x = this.x;
            vec.y = this.y;
            vec.direction = direction;
            this.trail.push(vec);
            this.direction = direction;
        }
    }

}
